ports-of-call/array.hpp

```diff
@@ -0,0 +1,262 @@
+#ifndef _PORTS_OF_CALL_ARRAY_HPP_
+#define _PORTS_OF_CALL_ARRAY_HPP_
+
+#include "portability.hpp"
+
+#include <cstddef>
+#include <iterator>
+#include <tuple>
+
+namespace PortsOfCall {
+
+//! The array class is a constexpr (mostly) replacement for std::array
+/*!
+std::array is not fully constexpr until C++20.  Currently we are relying on
+"constexpr"-ness to offload to GPUs.  PortsOfCall::array provides a
+constexpr replacement for std::array with the following notable exceptions:
+    - the reverse iterators are not provided because std::reverse_iterator is
+not constexpr until C++17 (could be provided with modest effort)
+
+    - the `at` function is not provided because it bounds checks and throws an
+exception which we cannot do on a GPU (it could be added and made
+non-constexpr)
+
+    - no std::get (cannot be provided without undefined behavior)
+*/
+
+namespace array_detail {
+
+template <typename T, std::size_t N>
+struct bits {
+  using type = T[N];
+
+  template <typename IntT>
+  PORTABLE_FUNCTION static constexpr T &ref_access(type &array,
```

Comment:
@dholladay00 should correct me if I'm wrong. Do we need `PORTABLE_FUNCTION` decorators for `constexpr`? Either way I think `PORTABLE_FORCEINLINE_FUNCTION` is better here than `PORTABLE_FUNCTION`.

Reply:
Several years ago we had some issues where `constexpr` wasn't completely sufficient.  It may have just been a quirk of an older compiler or something like that, and maybe they're redundant now.

Reply:
no I think the decorator may still be desirable... but since it's inline no reason not to use `PORTABLE_FORCEINLINE_FUNCTION` rather than `PORTABLE_FUNCTION`